numba/cuda/codegen.py

```diff
@@ -1,27 +1,186 @@
 from llvmlite import binding as ll
 from llvmlite import ir
 
+from numba.core import config, serialize
 from numba.core.codegen import Codegen, CodeLibrary
-from .cudadrv import nvvm
+from .cudadrv import devices, driver, nvvm
+
+import ctypes
+import numpy as np
+import os
+import subprocess
+import tempfile
 
 
 CUDA_TRIPLE = 'nvptx64-nvidia-cuda'
 
 
-class CUDACodeLibrary(CodeLibrary):
+def disassemble_cubin(cubin):
+    # nvdisasm only accepts input from a file, so we need to write out to a
+    # temp file and clean up afterwards.
+    fd = None
+    fname = None
+    try:
+        fd, fname = tempfile.mkstemp()
+        with open(fname, 'wb') as f:
+            f.write(cubin)
+
+        try:
+            cp = subprocess.run(['nvdisasm', fname], check=True,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
+        except FileNotFoundError as e:
+            if e.filename == 'nvdisasm':
+                msg = ("nvdisasm is required for SASS inspection, and has not "
+                       "been found.\n\nYou may need to install the CUDA "
+                       "toolkit and ensure that it is available on your "
+                       "PATH.\n")
+                raise RuntimeError(msg)
+        return cp.stdout.decode('utf-8')
+    finally:
+        if fd is not None:
+            os.close(fd)
+        if fname is not None:
+            os.unlink(fname)
+
 
-    def __init__(self, codegen, name):
+class CUDACodeLibrary(CodeLibrary, serialize.ReduceMixin):
```

Comment:
Although it won't affect the code here, mixins should go to the left. The MRO goes left to right. If `CodeLibrary` defines `__reduce__`, it will take priority over the one in the `ReduceMixin`.

Reply:
Many thanks - I've swapped these round now.